watertap/property_models/SOP_prop_pack.py

```diff
@@ -0,0 +1,551 @@
+        )
+
+        def rule_mass_frac_phase_comp(b, p, j):
+            return (
+                b.mass_frac_phase_comp[p, j] * b.flow_mass_phase[p]
+                == b.flow_mass_phase_comp[p, j]
+            )
+
+        self.eq_mass_frac_phase_comp = Constraint(
+            self.params.phase_list,
+            self.params.component_list,
+            rule=rule_mass_frac_phase_comp,
+        )
+
+    def _flow_vol_phase_comp(self):
+        self.flow_vol_phase_comp = Var(
+            self.params.phase_list,
+            self.params.component_list,
+            initialize=1e-6,
+            domain=NonNegativeReals,
+            units=units.m**3 / units.s,
+            doc="Component volumetric flowrate",
+        )
+
+        def rule_flow_vol_phase_comp(b, p, j):
+            return (
+                b.flow_vol_phase_comp[p, j] * b.dens_mass_phase_comp[p, j]
+                == b.flow_mass_phase_comp[p, j]
+            )
+
+        self.eq_flow_vol_phase_comp = Constraint(
+            self.params.phase_list,
+            self.params.component_list,
+            rule=rule_flow_vol_phase_comp,
+        )
+
+    def _flow_vol_phase(self):
+        self.flow_vol_phase = Var(
+            self.params.phase_list,
+            initialize=1e-3,
+            bounds=(None, None),
+            units=units.m**3 / units.s,
+            doc="Volumetric flow rate",
+        )
+
+        def rule_flow_vol_phase(b, p):
+            return b.flow_vol_phase[p] == sum(
+                b.flow_vol_phase_comp[p, j] for j in b.params.component_list
+            )
+
+        self.eq_flow_vol_phase = Constraint(
+            self.params.phase_list, rule=rule_flow_vol_phase
+        )
+
+    def _flow_mass_phase(self):
+        self.flow_mass_phase = Var(
+            self.params.phase_list,
+            initialize=1e-3,
+            bounds=(None, None),
+            units=units.kg / units.s,
+            doc="Mass flow rate",
+        )
+
+        def rule_flow_mass_phase(b, p):
+            return b.flow_mass_phase[p] == sum(
+                b.flow_mass_phase_comp[p, j] for j in b.params.component_list
+            )
+
+        self.eq_flow_mass_phase = Constraint(
+            self.params.phase_list, rule=rule_flow_mass_phase
+        )
+
+    def _dens_mass_phase(self):
+        self.dens_mass_phase = Var(
+            self.params.phase_list,
+            initialize=1e3,
+            bounds=(1e2, 2e3),
+            units=units.kg * units.m**-3,
+            doc="Mass density",
+        )
+
+        def rule_dens_mass_phase(b, p):
+            return b.dens_mass_phase[p] * b.flow_vol_phase[p] == b.flow_mass_phase[p]
+
+        self.eq_dens_mass_phase = Constraint(
+            self.params.phase_list, rule=rule_dens_mass_phase
+        )
+
+    def _vol_frac_phase_comp(self):
+        self.vol_frac_phase_comp = Var(
+            self.params.phase_list,
+            self.params.component_list,
+            initialize=0.1,
+            bounds=(None, None),
```

Comment:
Technically, bounds are 0 to 1, right?

Reply:
Sometimes you can get away with this and get better solver behaviour (e.g. mole fractions). However, you need to ensure there is an implicit upper bound in place (e.g., if the sum of the fractions must equal one and the lower bound is zero, then all fractions have an implicit upper bound of 1).

Reply:
I changed the bounds to be `(0, 1)`.